store/multistore/store.go

```diff
@@ -0,0 +1,301 @@
+package multistore
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"time"
+
+	"cosmossdk.io/log"
+	v1types "cosmossdk.io/store/types"
+	"cosmossdk.io/store/v2"
+	"cosmossdk.io/store/v2/commitment"
+	"github.com/cockroachdb/errors"
+	ics23 "github.com/cosmos/ics23/go"
+)
+
+type (
+	// MultiStore defines an abstraction layer containing a State Storage (SS) engine
+	// and one or more State Commitment (SC) engines.
+	//
+	// TODO:
+	// - Move relevant types to the 'core' package.
+	// - Remove reliance on store v1 types.
+	MultiStore interface {
+		GetSCStore(storeKey string) *commitment.Database
+		MountSCStore(storeKey string, sc *commitment.Database) error
+		GetProof(storeKey string, version uint64, key []byte) (*ics23.CommitmentProof, error)
+		LoadVersion(version uint64) error
+		LoadLatestVersion() error
+		GetLatestVersion() (uint64, error)
+		WorkingHash() []byte
+		Commit() ([]byte, error)
+		SetCommitHeader(h CommitHeader)
+
+		// TODO:
+		// - Tracing
+		// - Branching
+		// - Queries
+
+		io.Closer
+	}
+
+	CommitHeader interface {
+		GetTime() time.Time
+		GetHeight() uint64
+	}
+)
+
+var _ MultiStore = &Store{}
+
+type Store struct {
+	logger         log.Logger
+	commitHeader   CommitHeader
+	initialVersion uint64
+
+	// ss reflects the state storage backend
+	ss store.VersionedDatabase
+
+	// scStores reflect a mapping of store key to state commitment backend (i.e. a backend per module)
+	scStores map[string]*commitment.Database
+
+	// removalMap reflects module stores marked for removal
+	removalMap map[string]struct{}
+
+	// lastCommitInfo reflects the last version/hash that has been committed
+	lastCommitInfo *v1types.CommitInfo
+}
+
+func New(logger log.Logger, initialVersion uint64, ss store.VersionedDatabase) (MultiStore, error) {
+	return &Store{
+		logger:         logger.With("module", "multi_store"),
+		initialVersion: initialVersion,
+		ss:             ss,
+		scStores:       make(map[string]*commitment.Database),
+		removalMap:     make(map[string]struct{}),
+	}, nil
+}
+
+// Close closes the store and resets all internal fields. Note, Close() is NOT
+// idempotent and should only be called once.
+func (s *Store) Close() (err error) {
```

Comment:
Because this would make it idempotent which is not consistent with the rest of the store code. In fact, this was a comment you made previously (to make it not idempotent).